pyuvdata/fhd_cal.py

```diff
@@ -0,0 +1,195 @@
+            ra=obs_data['orig_phasera'][0], dec=obs_data['orig_phasedec'][0])
+        self.sky_catalog = cal_data['skymodel'][0]['catalog_name'][0]
+        self.ref_antenna_name = cal_data['ref_antenna_name'][0]
+        self.Nsources = int(cal_data['skymodel'][0]['n_sources'][0])
+        self.baseline_range = [float(cal_data['min_cal_baseline'][0]),
+                               float(cal_data['max_cal_baseline'][0])]
+
+        galaxy_model = cal_data['skymodel'][0]['galaxy_model'][0]
+        if galaxy_model == 0:
+            galaxy_model = None
+        else:
+            galaxy_model = 'gsm'
+
+        diffuse_model = cal_data['skymodel'][0]['diffuse_model'][0]
+        if diffuse_model == '':
+            diffuse_model = None
+        else:
+            diffuse_model = os.path.basename(diffuse_model)
+
+        if galaxy_model is not None:
+            if diffuse_model is not None:
+                self.diffuse_model = galaxy_model + ' + ' + diffuse_model
+            else:
+                self.diffuse_model = galaxy_model
+        elif diffuse_model is not None:
+            self.diffuse_model = diffuse_model
+
+        self.gain_convention = 'divide'
+        self.x_orientation = 'east'
+
+        self.set_gain()
+        fit_gain_array_in = cal_data['gain'][0]
+        fit_gain_array = np.zeros(self._gain_array.expected_shape(self), dtype=np.complex_)
+        for jones_i, arr in enumerate(fit_gain_array_in):
+            fit_gain_array[:, 0, :, 0, jones_i] = arr
+        if raw:
+            res_gain_array_in = cal_data['gain_residual'][0]
+            res_gain_array = np.zeros(self._gain_array.expected_shape(self), dtype=np.complex_)
+            for jones_i, arr in enumerate(fit_gain_array_in):
+                res_gain_array[:, 0, :, 0, jones_i] = arr
+            self.gain_array = fit_gain_array + res_gain_array
+        else:
+            self.gain_array = fit_gain_array
+
+        # FHD doesn't really have a chi^2 measure. What is has is a convergence measure.
+        # The solution converged well if this is less than the convergence
+        # threshold ('conv_thresh' in extra_keywords).
+        self.quality_array = np.zeros_like(self.gain_array, dtype=np.float)
+        convergence = cal_data['convergence'][0]
+        for jones_i, arr in enumerate(convergence):
+            self.quality_array[:, 0, :, 0, jones_i] = arr
+
+        # array of used frequencies (1: used, 0: flagged)
+        freq_use = obs_data['baseline_info'][0]['freq_use'][0]
+        # array of used antennas (1: used, 0: flagged)
+        ant_use = obs_data['baseline_info'][0]['tile_use'][0]
+        # array of used times (1: used, 0: flagged)
+        time_use = obs_data['baseline_info'][0]['time_use'][0]
+
+        time_array_use = time_array[np.where(time_use > 0)]
+        self.time_range = [np.min(time_array_use), np.max(time_array_use)]
+
+        # Currently this can't include the times because the flag array
```

Comment:
Should we make the format more flexible to allow for different-sized flag arrays? I'm not sure it's worth the effort since the solution here seems adequate, but if we think it is, we should make an issue.

Reply:
I'm of two minds on this one. The issue is that FHD has flag information per integration that we can't include because FHD generates a single calibration across all the times and the flag array is constrained to have the same shape as the calibration array.

I think that constraint makes some sense when you think about applying the calibration to data, but if you're trying to understand how the calibration was calculated you'd need all the flag info. Of course, there can also be visibility level flags that are already being lost and would affect the calibration solution, so maybe we're already in the regime where you don't have all the information.